transmat/zip/zip_unpack.go

```diff
@@ -0,0 +1,202 @@
+package ziptrans
+
+import (
+	"archive/zip"
+	"bytes"
+	"context"
+	"crypto/sha512"
+	"fmt"
+	"io"
+	"strings"
+	"time"
+
+	"github.com/polydawn/refmt/misc"
+	. "github.com/warpfork/go-errcat"
+
+	api "go.polydawn.net/go-timeless-api"
+	"go.polydawn.net/go-timeless-api/rio"
+	"go.polydawn.net/rio/fs"
+	"go.polydawn.net/rio/fsOp"
+	"go.polydawn.net/rio/lib/treewalk"
+	"go.polydawn.net/rio/transmat/mixins/buffer"
+	"go.polydawn.net/rio/transmat/mixins/filters"
+	"go.polydawn.net/rio/transmat/mixins/fshash"
+	"go.polydawn.net/rio/transmat/mixins/log"
+	"go.polydawn.net/rio/transmat/util"
+)
+
+func unpackZip(
+	ctx context.Context,
+	afs fs.FS,
+	filt api.FilesetUnpackFilter,
+	archiveWareID api.WareID,
+	reader io.Reader,
+	mon rio.Monitor,
+) (
+	prefilterWareID api.WareID,
+	actualWareID api.WareID,
+	err error,
+) {
+	defer RequireErrorHasCategory(&err, rio.ErrorCategory(""))
+
+	b := buffer.NewTemporaryBuffer()
+	defer b.Close()
+	readerAt, err := b.SectionReader(ctx, archiveWareID, reader, mon)
+	if err != nil {
+		return api.WareID{}, api.WareID{}, err
+	}
+
+	// Convert the raw byte reader to a zip stream.
+	zr, err := zip.NewReader(readerAt, readerAt.Size())
+	if err != nil {
+		return api.WareID{}, api.WareID{}, err
+	}
+
+	// Allocate bucket for keeping each metadata entry and content hash;
+	// the full tree hash will be computed from this at the end.
+	// We keep one for the raw ware data as we consume it, so we can verify no fuckery;
+	// we keep a second, separate one for the filtered data, which will compute a different hash.
+	prefilterBucket := &fshash.MemoryBucket{}
+	filteredBucket := &fshash.MemoryBucket{}
+
+	// Also allocate a map for keeping records of which dirs we've created.
+	dirs := map[fs.RelPath]struct{}{}
+
+	// Iterate over each entry, mutating filesystem as we go.
+	for _, f := range zr.File {
+		fmeta := fs.Metadata{}
+
+		// Check for done.
+		if err == io.EOF {
+			break // sucess!  end of archive.
+		}
+		if err != nil {
+			return api.WareID{}, api.WareID{}, Errorf(rio.ErrWareCorrupt, "corrupt zip: %s", err)
+		}
+		if ctx.Err() != nil {
+			return api.WareID{}, api.WareID{}, Errorf(rio.ErrCancelled, "cancelled")
+		}
+
+		// Reshuffle metainfo to our default format.
+		skipMe, haltMe := ZipHdrToMetadata(&f.FileHeader, &fmeta)
+		if skipMe != nil {
+			// n.b. every time this happens in practice so far, it's a `g` header,
```

Comment:
Also comment freshness check ;)

It looks like `ZipHdrToMetadata` doesn't ever have a branch that returns `skipMe` at all.  So istm you could just remove that from that function's returns, and drop this branch here entirely.  (Though there's some often-useful convention of packages like this having a `FoofmtHdrToMetadata` function, there's no actual interface ZipHdrToMetadata is conforming to... it can definitely drop returns and have a different signature than its siblings-by-convention.)